File: src/main/kotlin/codeacademy/bookingforum/app/topic/TopicService.java
package codeacademy.bookingforum.app.topic;

import codeacademy.bookingforum.app.configuration.ResponseObject;
import codeacademy.bookingforum.app.ecxeption.global.UnsatisfiedExpectationException;
import codeacademy.bookingforum.app.topicCategory.TopicCategory;
import codeacademy.bookingforum.app.topicCategory.TopicCategoryRepo;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.context.request.WebRequest;

import java.util.Collections;
import java.util.List;

@Service
public class TopicService {
    @Autowired
    TopicRepository topicRepo;
    @Autowired
    TopicMapper topicMapper;
    @Autowired
    TopicCategoryRepo sectionRepo;

    public ResponseObject create(TopicDto topic, WebRequest request) {
        TopicCategory section = sectionRepo.findById(topic.getSectionId()).orElse(null);
        if (section == null) {
            throw new UnsatisfiedExpectationException("Can't save topic to non-existing section!");
        }
        Topic oldTopic = topicRepo.findByTitle(topic.getTitle().trim());
        if (oldTopic != null) {
            throw new UnsatisfiedExpectationException("This topic already exists!");
        }

        section.getTopics().add(topicRepo.save(topicMapper.fromDto(topic)));

        return new ResponseObject(Collections.singletonList("Successfully added new topic."), HttpStatus.CREATED, request);
    }

    public List<TopicDto> getList(Long id) {
        TopicCategory section = sectionRepo.findById(id).orElse(null);
        if (section == null) {
            throw new UnsatisfiedExpectationException("Section with id "+id+" does not exist!");
        }

        List<Topic> topics = topicRepo.findAllBySection(section);
        return topicMapper.toDtoList(topics);
    }
}

File: src/main/kotlin/codeacademy/bookingforum/app/topic/TopicService.kt
package codeacademy.bookingforum.app.topic

import codeacademy.bookingforum.app.topicCategory.TopicCategory
import codeacademy.bookingforum.app.user.auth.UserAuth
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service

@Service
class TopicService {

    @Autowired
    private lateinit var mapper: TopicMapper

    @Autowired
    private lateinit var repository: TopicRepository
    fun getAllTopics(): List<TopicDto?> {
        return repository.findAll().map { mapper.toDto(it) }
    }

    fun getTopicById(id: Long): TopicDto? {
        return repository.findByIdOrNull(id).let { mapper.toDto(it) }
    }

    fun postTopic(topicDto: TopicDto?) {
        mapper.fromDto(topicDto)?.let { repository.save(it) }
    }

    fun updateTopic(id: Long, topicDto: TopicDto?): String {
        return if (repository.existsById(id)) {
            repository.save(
                Topic(
                    id,
                    topicDto?.title,
                    topicDto?.description,
                    topicDto?.postCount,
                    topicDto?.replayCount,
                    topicDto?.postLast,
                    topicDto?.userLastPost,
                    UserAuth(topicDto?.userId),
                    TopicCategory(topicDto?.topicCategory_id)
                )
            )
            "Topic with Id $id updated"
        } else {
            "Topic with $id does not exist"
        }
    }

    fun deleteTopic(id: Long): String {
        return if (repository.existsById(id)) {
            repository.deleteById(id)
            "Deleted topic with id $id"
        } else {
            "Topic with $id not found"
        }
    }
}